File: CSharp/NetCubeSolver/PuzzleSolution.cs
using System.Text;
using NetCubeSolver.ExactCover;

namespace NetCubeSolver;

public class PuzzleSolution
{
    
    private readonly int _size;
    private readonly Dictionary<int, string> _puzzlePieceSymbols;
    
    private int [,,] _cubes;

    public PuzzleSolution(int size, Dictionary<int, string> puzzlePieceSymbols)
    {
        _size = size;
        _puzzlePieceSymbols = puzzlePieceSymbols;
        InitEmptyCubes();
    }

    public List<int> GetPositionHeaders()
    {
        return Enumerable.Range(0, _size*_size*_size).ToList();
    }

    public static PuzzleSolution ConvertToSolution(int size, Dictionary<int, string> puzzlePieceSymbols, int zDiv, List<int> solutionData)
    {
        var solution = new PuzzleSolution(size, puzzlePieceSymbols);
        List<int> pieceData = new List<int>();
        int currentPieceId = -1;
        int pieceCount = puzzlePieceSymbols.Count - 1;
        foreach (var point in solutionData)
        {
            if (point != -1)
            {
                if (puzzlePieceSymbols.ContainsKey(point))
                {
                    currentPieceId = point;
                }
                else
                {
                    pieceData.Add(point - pieceCount);
                }
            }
            else
            {
                solution.FillFromPieceData(currentPieceId, pieceData, zDiv);
                pieceData.Clear();
            }
        }
        
        return solution;
    }

    private void FillFromPieceData(int pieceId, List<int> pieceData, int zDiv)
    {
        foreach (var point in pieceData)
        {
            int x = point;
            int z = Math.DivRem(x, zDiv, out x);
            int y = Math.DivRem(x, _size, out x);
            
            _cubes[z, y, x] = pieceId;
            
        }
    }

    // The canonical solution is the one with the lowest string representation
    public string FindAndSetToCanonicalSolution()
    {
        int orientation = 0;
        Dictionary<string, int[,,]> orientations = new Dictionary<string, int[,,]>();
        StringBuilder sb = new StringBuilder(_size * _size * _size);
        
        orientations.Add(GetCanonicalString(_cubes, sb), _cubes);
        sb.Clear();
        
        var cubes = _cubes;
        for (int i = 0; i < 23; i++)
        {
            cubes = Rotate(i, cubes);
            orientations.Add(GetCanonicalString(cubes, sb), cubes);
            sb.Clear();
        }
        
        var keys = orientations.Keys.ToArray();
        Array.Sort(keys);
        _cubes = orientations[keys[0]];
        return keys[0];
        
    }

    private string GetCanonicalString(int[,,] cubes, StringBuilder sb)
    {
        for (int z = 0; z < _size; z++)
        {
            for (int y = 0; y < _size; y++)
            {
                for (int x = 0; x < _size; x++)
                {
                    sb.Append(_puzzlePieceSymbols[cubes[z,y,x]]);
                }
            }
        }
        return sb.ToString();
    }

    public string RenderToString(string title)
    {
        StringBuilder sb = new StringBuilder();
        sb.AppendLine(title);
        sb.AppendLine();
        for (int y = 0; y < _size; y++)
        {
            string line = "\t";
            for (int z = 0; z < _size; z++)
            {
                for (int x = 0; x < _size; x++)
                {
                    line += $"{_puzzlePieceSymbols[_cubes[z,y,x]]} ";
                }

                line += "  ";
            }
            sb.AppendLine(line);
        }
        sb.AppendLine();
        return sb.ToString();
    }
    
    public void WriteToTextFile(string title, StreamWriter writer)
    {
        writer.WriteLine(title);
        writer.WriteLine();
        for (int y = 0; y < _size; y++)
        {
            writer.Write("\t");
            for (int z = 0; z < _size; z++)
            {
                for (int x = 0; x < _size; x++)
                {
                    writer.Write($"{_puzzlePieceSymbols[_cubes[z,y,x]]} ");
                }

                writer.Write("  ");
            }
            writer.WriteLine();
        }
        writer.WriteLine();
    }

    public void Rotate(Axis axis)
    {
        _cubes = Rotate(axis, _cubes);
    }

    public bool RemovePiece(int pieceId)
    {
        bool ret = false;
        for (int z = 0; z < _size; z++)
        {
            for (int y = 0; y < _size; y++)
            {
                for (int x = 0; x < _size; x++)
                {
                    if (_cubes[z,y,x] == pieceId)
                    {
                        _cubes[z,y,x] = -1;
                        ret = true;
                    }
                }
            }
        }

        return ret;
    }

    private bool DoInsert(PuzzlePiece piece, bool permanent)
    {
        bool ret = true;
        foreach (var cube in piece.Geometry)
        {
            if (_cubes[cube.Z, cube.Y, cube.X] == -1)
            {
                if (permanent)
                {
                    _cubes[cube.Z, cube.Y, cube.X] = piece.Id;
                }
            }
            else
            {
                ret = false;
            }
        }

        if (!ret && permanent) RemovePiece(piece.Id);
        
        return ret;
    }

    public bool InsertPiece(PuzzlePiece piece)
    {
        bool ret = DoInsert(piece, true);
        if (!ret) return false;
        
        return ret;
    }
    
    private int[,,] CreateEmptyCubes()
    {
        var cubes = new int[_size,_size,_size];
        for (int z = 0; z < _size; z++)
        {
            for (int y = 0; y < _size; y++)
            {
                for (int x = 0; x < _size; x++)
                {
                    cubes[z,y,x] = -1;
                }
            }
        }
        return cubes;
    }

    private void InitEmptyCubes()
    {
        _cubes = CreateEmptyCubes();
    }
    
    private int[,,] Rotate(Axis axis, int[,,] currentCubes)
    {
        var cubes = new int[_size,_size,_size];

        for (int z = 0; z < _size; z++)
        {
            for (int y = 0; y < _size; y++)
            {
                for (int x = 0; x < _size; x++)
                {
                    switch (axis)
                    {
                        case Axis.X:
                            cubes[(_size - 1) - y,z,x] = currentCubes[z,y,x];
                            break;
                        case Axis.Y:
                            cubes[x,y,(_size - 1)- z] = currentCubes[z,y,x];
                            break;
                        case Axis.Z:
                            cubes[z, (_size - 1) - x, y] = currentCubes[z,y,x];
                            break;
                    }
                }
            }
        }
        return cubes;
    }
    
    private int[,,] Rotate(int currentOrientation, int[,,] currentCubes)
    {
        var cubes = currentCubes;
        var newOrientation = currentOrientation +1;
        cubes = Rotate(Axis.X, cubes);
        switch (newOrientation)
        {
            case 4:
            case 8:
            case 12:
                cubes = Rotate(Axis.Y, cubes);
                break;
            case 16:
                cubes = Rotate(Axis.Y, cubes);
                cubes = Rotate(Axis.Z, cubes);
                break;
            case 20:
                cubes = Rotate(Axis.Z, cubes);
                cubes = Rotate(Axis.Z, cubes);
                break;
        }

        return cubes;
    }
    
}

File: Kotlin/app/src/main/kotlin/ca/jwolf/cubesolver/datamodels/PuzzleSolution.kt
package ca.jwolf.cubesolver.datamodels

import ca.jwolf.cubesolver.configs.Axis
import ca.jwolf.cubesolver.configs.PuzzlePiece
import java.io.BufferedWriter

class PuzzleSolution(puzzleSize: Int, pieceSymbolsLookup: Map<Int, String>) {
    val size: Int = puzzleSize
    private val puzzlePieceSymbolsLookup: Map<Int, String> = pieceSymbolsLookup
    private var cubes: Array<Array<Array<Int>>> = emptyArray()

    init {
        cubes = createEmptyCubes()
    }

    constructor(puzzleSize: Int, pieceSymbolsLookup: Map<Int, String>, zDiv: Int, solutionData: List<Int>) : this(puzzleSize, pieceSymbolsLookup) {

        val pieceData: MutableList<Int> = mutableListOf()
        var currentPieceId = -1
        val pieceCount = puzzlePieceSymbolsLookup.size - 1

        // solutionData is a stream of ints representing the solved cube.
        // it can be divided in sections with each section separated by the value of -1
        // each value is a column from the cover matrix that represents that column being 'true' for that row
        // if the column value is within the piece lookup, then it represents what piece the current section is for
        // otherwise it represents a position within the solution that the piece occupies.

        solutionData.forEach { point ->
            // -1 is a delimiter between different pieces in the stream
            if (point != -1) {
                if (puzzlePieceSymbolsLookup.containsKey(point))
                {
                    currentPieceId = point
                }
                else
                {
                    pieceData.add(point - pieceCount)
                }
            }
            else
            {
                fillFromPieceData(currentPieceId, pieceData, zDiv)
                pieceData.clear()
            }
        }

    }

    fun getPositionalHeaders(): List<Int> {
        return (0 until size*size*size).map { it }
    }

    fun insertPiece(piece: PuzzlePiece): Boolean {
        var ret = true
        piece.geometry().forEach{ component ->
            if(cubes[component.z][component.y][component.x] == -1)
            {
                cubes[component.z][component.y][component.x] = piece.id
            }
            else
            {
                ret = false

            }
        }

        if (!ret) {
            removePiece(piece.id)
        }
        return ret
    }

    fun removePiece(pieceId: Int) {
        for (z in 0 until size) {
            for (y in 0 until size) {
                for (x in 0 until size) {
                    if (cubes[z][y][x] == pieceId) {
                        cubes[z][y][x] = -1
                    }
                }
            }
        }
    }

    fun renderToString(title: String): String {
        val sb = StringBuilder()
        sb.appendLine(title)
        sb.appendLine()
        for (y in 0 until size) {
            var line = "\t"
            for (z in 0 until size) {
                for (x in 0 until size) {
                    line = line.plus("${puzzlePieceSymbolsLookup[cubes[z][y][x]]} ")
                }
                line = line.plus("  ")
            }
            sb.appendLine(line)
        }
        sb.appendLine()
        return sb.toString()
    }

    fun getCanonicalSolutionString(): String {
        val orientations: MutableMap<String, Array<Array<Array<Int>>>> = mutableMapOf()
        val sb = StringBuilder()
        orientations[getCanonicalString(cubes, sb)] = cubes
        sb.clear()

        var tempCubes = cubes
        for (i in 0 until 23)
        {
            tempCubes = rotate(i, tempCubes)
            orientations[getCanonicalString(tempCubes, sb)] = tempCubes
            sb.clear()
        }

        val sortedKeys = orientations.keys.sorted()
        cubes = orientations[sortedKeys[0]]!!
        return sortedKeys[0]

    }

    fun writeToTextFile(title: String, writer: BufferedWriter) {
        writer.write(title)
        writer.newLine()
        writer.newLine()
        for (y in 0 until size) {
            writer.write("\t")
            for (z in 0 until size) {
                for (x in 0 until size) {
                    writer.write("${puzzlePieceSymbolsLookup[cubes[z][y][x]]} ")
                }
                writer.write("  ")
            }
            writer.newLine()
        }
        writer.newLine()
    }

    private fun getCanonicalString(cubes: Array<Array<Array<Int>>>, sb: StringBuilder): String {
        for (z in 0 until size) {
            for (y in 0 until size) {
                for (x in 0 until size) {
                    sb.append(puzzlePieceSymbolsLookup[cubes[z][y][x]])
                }
            }
        }
        return sb.toString()
    }

    private fun createEmptyCubes(): Array<Array<Array<Int>>> {
        return Array(size) { Array(size) { Array(size) { -1 } } }
    }

    private fun fillFromPieceData(pieceId: Int, pieceData: List<Int>, zDiv: Int) {
        pieceData.forEach { point ->
            var x = point
            val z = x / zDiv
            x %= zDiv
            val y = x / size
            x %= size

            cubes[z][y][x] = pieceId
        }
    }

    private fun rotate(axis: Axis, currentCubes: Array<Array<Array<Int>>>): Array<Array<Array<Int>>> {
        var tempCubes = createEmptyCubes()
        for (z in 0 until size) {
            for (y in 0 until size) {
                for (x in 0 until size) {
                    when(axis){
                        Axis.X -> tempCubes[(size - 1) - y][z][x] = currentCubes[z][y][x]
                        Axis.Y -> tempCubes[x][y][(size - 1) - z] = currentCubes[z][y][x]
                        Axis.Z -> tempCubes[z][(size - 1) - x][y] = currentCubes[z][y][x]
                    }
                }
            }
        }
        return tempCubes
    }

    private fun rotate(currentOrientation: Int, currentCubes: Array<Array<Array<Int>>>): Array<Array<Array<Int>>> {
        var tempCubes = currentCubes
        val newOrientation = currentOrientation + 1
        tempCubes = rotate(Axis.X, tempCubes)
        when(newOrientation){
            4,8,12 -> tempCubes = rotate(Axis.Y, tempCubes)
            16 -> {
                tempCubes = rotate(Axis.Y, tempCubes)
                tempCubes = rotate(Axis.Z, tempCubes)
            }
            20 -> {
                tempCubes = rotate(Axis.Z, tempCubes)
                tempCubes = rotate(Axis.Z, tempCubes)
            }
        }
        return tempCubes
    }


}